magda-typescript-common/src/express/status.ts

```diff
@@ -0,0 +1,204 @@
+import { Router } from "express";
+import request = require("request");
+
+export type Probe = () => Promise<State>;
+
+export interface ProbesList {
+    [id: string]: Probe;
+}
+
+export interface OptionsIO {
+    // inputs
+    probes?: ProbesList;
+    probeUpdateMs?: number;
+    forceRun?: boolean;
+    // outputs
+    ready?: boolean;
+    _installed?: any;
+    since?: string;
+    last?: string;
+    details?: any;
+    error?: any;
+}
+
+export interface State {
+    ready: boolean;
+    since?: string;
+    last?: string;
+    details?: any;
+    error?: any;
+    latencyMs?: number;
+}
+
+export function createStatusRouter(options: OptionsIO = {}): Router {
+    const router = Router();
+
+    installUpdater(options);
+
+    // we don't want to cache status
+    router.use(function(req, res, next) {
+        res.set("Cache-Control", "no-cache");
+        next();
+    });
+
+    // liveness probe for knowing if service is up
+    router.get("/live", function(req, res) {
+        res.status(200).send("OK");
+    });
+
+    // readiness probe for knowing if service and its dependents are
+    // ready to go
+    router.get("/ready", function(req, res) {
+        res.status(options.ready ? 200 : 500).json({
+            ready: options.ready,
+            since: options.since,
+            last: options.last,
+            details: options.details,
+            now: new Date().toISOString()
+        });
+    });
+
+    return router;
+}
+
+function installUpdater(options: OptionsIO) {
+    // This is so that we cab share OptionsIO objects
+    // for multiple endpoints
+    if (!options._installed) {
+        options.ready = !options.probes;
+        options.probeUpdateMs = options.probeUpdateMs || 10000;
+        options.details = {};
+        options.since = new Date().toISOString();
+        options._installed = {};
+    }
+    if (!options.probes) {
+        return;
+    }
+    // every probe deserves to not be blocked by state of other probes
+    // so we have parallel probe checkers
+    Object.entries(options.probes).forEach(probe => {
+        const [id, callback] = probe;
```

Comment:
Pulling out readiness/liveness checks into their own router is a great idea, and I really like the idea of just passing async functions that check the status.

This bit I don't understand though - it looks a lot like a whole new readiness/liveness check system? What is the advantage of doing this in our code vs letting Kubernetes do it?

I'm worried because:
- This is really crucial, and we're now doing it via code that we have to maintain, rather than Kubernetes code that's used and tested by thousands of people.
- It removes the ability to configure the initial delay/cadence of readiness/liveness checks - e.g. some services require a bit more time to start, and this changes from environment to environment
- It's not obvious that the options in Kubernetes don't work anymore... also they do still work for some pods (e.g. scala ones) but not for others. This means that someone might set something to be checked every 5 seconds and restart after 5 failures - this will only translate to 1 actual check via this system, resulting in k8s killing the pod after only one failed check.
- Currently if a service fails due to liveness checks failing, `kubectl describe pod <pod>` will give you the details of how many failed, but now that information is going to be false because the actual checks will be done inside the pod

Reply:
Re: our verbal conversation, obviously I did _not_ understand how this all worked 😂 . But it'd still be good to change the readiness check to something more like the liveness one so that any service failure doesn't take down the gateway and keep this for monitoring checks.

Reply:
Oh and maybe a comment to explain that this doesn't replace all the readiness checks, just gives us our own view!